File: framework/src/main/java/com/linxiao/framework/list/ViewBindingRecyclerHolder.java
package com.linxiao.framework.list;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.viewbinding.ViewBinding;

import java.lang.reflect.InvocationTargetException;
import java.util.Objects;

/**
 * <p>
 * RecyclerViewHolder built with ViewBinding
 * the constructor method is private and use {@link #create(Class, ViewGroup)} method
 * to create ViewHolder.
 * use holder.getViewBinding.someView to get reference of View
 * </p>
 *
 * @author linxiao
 * @since 2020-10-20
 */
public class ViewBindingRecyclerHolder<B extends ViewBinding> extends BaseRecyclerViewHolder {

    private B binding = null;

    public static <B extends ViewBinding> ViewBindingRecyclerHolder<B> create(Class<B> bindingClass, ViewGroup parent) {
        return new ViewBindingRecyclerHolder<>(bindingClass, parent);
    }

    protected ViewBindingRecyclerHolder(Class<B> bindingClass, ViewGroup parent) {
        this(inflateItemView(bindingClass, parent));
        initViewBinding(bindingClass);
    }

    private ViewBindingRecyclerHolder(View itemView) {
        super(itemView);
    }

    public B getViewBinding() {
        return binding;
    }

    @SuppressWarnings("unchecked")
    private void initViewBinding(Class<B> bindingClass) {
        try {
            binding = (B) bindingClass.getMethod("bind", View.class).invoke(null, itemView);
        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("unchecked")
    private static <B extends ViewBinding> View inflateItemView(Class<B> bindingClass, ViewGroup parent) {
        B binding = null;
        try {
            binding = (B) bindingClass.getMethod("inflate", LayoutInflater.class, ViewGroup.class, boolean.class)
                    .invoke(null, LayoutInflater.from(parent.getContext()), parent, false);
        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            e.printStackTrace();
        }
        return Objects.requireNonNull(binding).getRoot();
    }
}

File: framework/src/main/java/com/linxiao/framework/list/ViewBindingRecyclerHolder.kt
package com.linxiao.framework.list

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import java.lang.reflect.InvocationTargetException
import java.util.Objects

/**
 *
 *
 * RecyclerViewHolder built with ViewBinding
 * the constructor method is private and use [.create] method
 * to create ViewHolder.
 * use holder.getViewBinding.someView to get reference of View
 *
 *
 * @author linxiao
 * @since 2020-10-20
 */
class ViewBindingRecyclerHolder<B : ViewBinding> constructor(
    bindingClass: Class<B>,
    parent: ViewGroup
) : RecyclerView.ViewHolder(inflateItemView(bindingClass, parent)) {

    val viewBinding by lazy {
        initViewBinding(bindingClass)
    }

    @Suppress("UNCHECKED_CAST")
    private fun initViewBinding(bindingClass: Class<B>): B {
        return bindingClass.getMethod("bind", View::class.java).invoke(null, itemView) as B
    }

    companion object {
        fun <B : ViewBinding> create(
            bindingClass: Class<B>,
            parent: ViewGroup
        ): ViewBindingRecyclerHolder<B> {
            return ViewBindingRecyclerHolder(bindingClass, parent)
        }

        @Suppress("UNCHECKED_CAST")
        private fun <B : ViewBinding?> inflateItemView(
            bindingClass: Class<B>,
            parent: ViewGroup
        ): View {
            var binding: B? = null
            try {
                binding = bindingClass.getMethod(
                    "inflate",
                    LayoutInflater::class.java,
                    ViewGroup::class.java,
                    Boolean::class.javaPrimitiveType
                ).invoke(null, LayoutInflater.from(parent.context), parent, false) as B
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            }
            return binding!!.root
        }
    }
}